demo_modules/wind/wind.js

```diff
@@ -51,55 +71,74 @@ class WindClient extends ModuleInterface.Client {
   finishFadeOut() {
     this.mapSurface.destroy();
     this.overlaySurface.destroy();
+    this.animationSurface.destroy();
   }
 
   willBeShownSoon(container, deadline) {
     const CanvasSurface = require('client/surface/canvas_surface');
     this.mapSurface = new CanvasSurface(container, wallGeometry);
+    this.mapSurface.pushOffset();
     this.overlaySurface = new CanvasSurface(container, wallGeometry);
+    this.overlaySurface.pushOffset();
+    this.animationSurface = new CanvasSurface(container, wallGeometry);
+    this.animationSurface.pushOffset();
+
+    this.radius = globalWallGeometry.extents.w / 3;
+    this.scale = 2.5*this.radius;
 
     this.projection = d3.geoOrthographic()
-      .scale(diameter/2.1)
-      .rotate([100, -400])
-      .translate([canvasWidth/2, canvasHeight/2])
+      .scale(this.scale)
+      .rotate([ROTATEX, ROTATEY])
+      .translate([globalWallGeometry.extents.w/2, globalWallGeometry.extents.h/2])
       .clipAngle(90);
 
+    this.bounds = clampedBounds(
+        d3.geoPath().projection(this.projection).bounds({type: "Sphere"}),
+        globalWallGeometry.extents.w, globalWallGeometry.extents.h);
+
     this.mapLoaded = loadJson('americas.json').then((mapData) => {
       this.mapData = mapData;
     });
 
-    this.forecastLoaded = loadJson('current-wind-surface-level-gfs-1.0.json')
+    this.dataProcessed = loadJson('current-wind-surface-level-gfs-1.0.json')
       .then((file) => {
         this.grid = new ForecastGrid(file);
-        this.field = VectorField.create(this.projection, this.grid);
+        this.vectorField = VectorField.create(this.projection, this.bounds,
+            this.grid);
+      }).then(() => {
+        this.particleField = new ParticleField(this.bounds,
+            this.grid, this.vectorField, this.animationSurface.context);
       });
-    return Promise.all([this.mapLoaded, this.forecastLoaded]);
+    return Promise.all([this.mapLoaded, this.dataProcessed]);
   }
 
   draw(time, delta) {
-    if (!this.drawn) {
-      this.drawn = true;
-      this.mapLoaded.then(() => {
-        return this.drawMap(this.projection, this.mapSurface.context,
-            this.mapData);
-      });
-      this.forecastLoaded.then(() => {
-        return this.drawOverlay(this.overlaySurface.context, this.field);
-      });
+    if (this.particleField) {
+      if (!this.mapDrawn) {
+        this.drawMap(this.projection, this.mapSurface.context, this.mapData);
+        this.drawOverlay(this.overlaySurface.context, this.vectorField);
+        this.particleField.draw();
```

Comment:
Like here... won't this copy xt to x for no reason?

Reply:
ACK